# **Projeto de exploração de dados de vendas de carros do eBay**

Prof. Ivanovitch Silva

UFRN - 2021.2

## **Integrantes**

*   Kathleen Noemi Duarte Rego

*   Pedro Henrique de Souza Fonsêca dos Santos

## **Informações do Projeto**

### O dataset é composto de 50.000 dados de vendas e nosso objetivo é limpar os dados e analisar as listagens de carros.



A base é composta pelos dados a seguir:

* **dateCrawled:** Quando este anúncio foi rastreado pela primeira vez. Todos os valores de campo são obtidos a partir desta data;
* **name:** Nome do carro;
* **seller:** Se o vendedor é particular ou revendedor.
* **offerType:** O tipo de listagem
* **price:** O preço do anúncio de venda do carro.
* **abtest:** Se a lista está incluída em um teste A / B.
* **vehicleType:** O tipo de veículo.
* **yearOfRegistration:** O ano em que o carro foi registrado pela primeira vez.
* **gearbox:** O tipo de transmissão.
* **powerPS:** O poder do carro no PS.
* **model:** O nome do modelo do carro.
* **kilometer:** Quantos quilômetros o carro percorreu.
* **monthOfRegistration:** O mês em que o carro foi registrado pela primeira vez.
* **fuelType:** Que tipo de combustível o carro usa.
* **brand:** A marca do carro.
* **notRepairedDamage:** Se o carro apresentar danos que ainda não foram reparados.
* **dateCreated:** A data em que a lista do eBay foi criada.
* **nrOfPictures:** O número de fotos no anúncio.
* **postalCode:** O código postal da localização do veículo.
* **lastSeenOnline:** Quando o rastreador viu este anúncio online pela última vez.

In [4]:
import pandas as pd

autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [5]:
autos

dateCrawled  ...             lastSeen
0      2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1      2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2      2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3      2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4      2016-04-01 14:38:50  ...  2016-04-01 14:38:50
...                    ...  ...                  ...
49995  2016-03-27 14:38:19  ...  2016-04-01 13:47:40
49996  2016-03-28 10:50:25  ...  2016-04-02 14:18:02
49997  2016-04-02 14:44:48  ...  2016-04-04 11:47:27
49998  2016-03-08 19:25:42  ...  2016-04-05 16:45:07
49999  2016-03-14 00:42:12  ...  2016-04-06 21:18:48

[50000 rows x 20 columns]

In [6]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled  ...             lastSeen
0  2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1  2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2  2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3  2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4  2016-04-01 14:38:50  ...  2016-04-01 14:38:50

[5 rows x 20 columns]

Podemos ver que nas colunas vehicleType, gearbox, model, fuelType, notRepairedDamage há valores nulos que precisam ser limpados.

### Convertendo o padrão dos nomes das colunas

O padrão dos nomes das colunas estão em camelcase, e mudaremos para snakecase

In [7]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [8]:
mapping_dict = {'yearOfRegistration': 'registration_year',
                'monthOfRegistration': 'registration_month',
                'notRepairedDamage': 'unrepaired_damage',
                'dateCreated': 'ad_created',
                'postalCode': 'postal_code',
                'lastSeen': 'last_seen',
                'nrOfPictures': 'number_of_pictures',
                'vehicleType': 'vehicle_type',
                'fuelType': 'fuel_type',
                'dateCrawled': 'ad_crawled',
                'offerType': 'offer_type',
                'powerPS': 'power_ps'}
autos.rename(columns = mapping_dict, inplace=True)
autos.head()

ad_crawled  ...            last_seen
0  2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1  2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2  2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3  2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4  2016-04-01 14:38:50  ...  2016-04-01 14:38:50

[5 rows x 20 columns]

As colunas foram modificadas para o padrão snakecase, que utiliza o separador '_'.

### Explorando dados que podem ser modificados

Verificaremos os dados que poderiam se tornar numéricos, que poderiam ser retirados do dataset ou que estão com valores considerados estranhos.

In [9]:
autos.describe(include='all')

ad_crawled         name  ...   postal_code            last_seen
count                 50000        50000  ...  50000.000000                50000
unique                48213        38754  ...           NaN                39481
top     2016-03-23 18:39:34  Ford_Fiesta  ...           NaN  2016-04-07 06:17:27
freq                      3           78  ...           NaN                    8
mean                    NaN          NaN  ...  50813.627300                  NaN
std                     NaN          NaN  ...  25779.747957                  NaN
min                     NaN          NaN  ...   1067.000000                  NaN
25%                     NaN          NaN  ...  30451.000000                  NaN
50%                     NaN          NaN  ...  49577.000000                  NaN
75%                     NaN          NaN  ...  71540.000000                  NaN
max                     NaN          NaN  ...  99998.000000                  NaN

[11 rows x 20 columns]

Temos que:
Colunas com valores estranhos: number_of_pictures, registration_year

Colunas com valores redundantes: offer_type e seller

Colunas com valores que poderiam se tornar numéricos: odometer, price

Agora iremos transformar os dados de odômetro e preço em valores numéricos.

In [10]:
autos['price'] = autos['price'].str.replace('$','').str.replace(',','').astype(int)
autos['odometer'] = autos['odometer'].str.replace('km','').str.replace(',','').astype(int)
autos.rename(columns = {'odometer' : 'odometer_km'}, inplace = True)

### Retirando valores outliers

Outliers são valores que se diferenciam demais do resto dos valores e que levanta suspeitas de sua veracidade. Iremos retira-los.

In [11]:
autos.head()

ad_crawled  ...            last_seen
0  2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1  2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2  2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3  2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4  2016-04-01 14:38:50  ...  2016-04-01 14:38:50

[5 rows x 20 columns]

In [12]:
autos['odometer_km'].describe()
autos['odometer_km'].value_counts().sort_index(ascending=True)

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

In [13]:
autos['price'].describe()
autos['price'].value_counts().sort_index(ascending=True).tail(50)

98500       1
99000       2
99900       2
104900      1
105000      2
109999      1
114400      1
115000      1
115991      1
116000      1
119500      1
119900      1
120000      2
128000      1
129000      1
130000      1
135000      1
137999      1
139997      1
145000      1
151990      1
155000      1
163500      1
163991      1
169000      1
169999      1
175000      1
180000      1
190000      1
194000      1
197000      1
198000      1
220000      1
250000      1
259000      1
265000      1
295000      1
299000      1
345000      1
350000      1
999990      1
999999      2
1234566     1
1300000     1
3890000     1
10000000    1
11111111    2
12345678    3
27322222    1
99999999    1
Name: price, dtype: int64

Não foi identificado um outliers considerável no odômetro, mas no preço sim. Logo, removemos os valores que eram menores que 500 e maiores que 100.000.

In [14]:
autos = autos[autos['price'].between(500,100000)]

### Calculando distribuição de valores por data

Nesse ponto iremos utilizar as datas para criar uma distribuição que nos dirá a porcentagem de vendas em cada dia, utilizando as colunas que contém datas.

In [15]:
autos['ad_crawled'].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.025545
2016-03-06    0.014137
2016-03-07    0.036220
2016-03-08    0.033157
2016-03-09    0.032913
2016-03-10    0.032736
2016-03-11    0.033046
2016-03-12    0.037330
2016-03-13    0.015536
2016-03-14    0.036331
2016-03-15    0.034001
2016-03-16    0.029362
2016-03-17    0.031160
2016-03-18    0.012850
2016-03-19    0.034777
2016-03-20    0.038084
2016-03-21    0.037729
2016-03-22    0.032935
2016-03-23    0.032403
2016-03-24    0.028985
2016-03-25    0.031093
2016-03-26    0.032625
2016-03-27    0.031160
2016-03-28    0.034822
2016-03-29    0.033290
2016-03-30    0.033290
2016-03-31    0.031670
2016-04-01    0.033867
2016-04-02    0.035798
2016-04-03    0.038817
2016-04-04    0.036642
2016-04-05    0.013161
2016-04-06    0.003174
2016-04-07    0.001354
Name: ad_crawled, dtype: float64

Distribuição de valores da colunas ad_crawled

In [16]:
autos['ad_created'].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2015-06-11    0.000022
2015-08-10    0.000022
2015-09-09    0.000022
2015-11-10    0.000022
2015-12-05    0.000022
                ...   
2016-04-03    0.039039
2016-04-04    0.037019
2016-04-05    0.011896
2016-04-06    0.003262
2016-04-07    0.001198
Name: ad_created, Length: 76, dtype: float64

Distribuição de valores da colunas ad_created

In [17]:
autos['last_seen'].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.001087
2016-03-06    0.004172
2016-03-07    0.005216
2016-03-08    0.007013
2016-03-09    0.009454
2016-03-10    0.010298
2016-03-11    0.012051
2016-03-12    0.023925
2016-03-13    0.008877
2016-03-14    0.012295
2016-03-15    0.015691
2016-03-16    0.016157
2016-03-17    0.027698
2016-03-18    0.007390
2016-03-19    0.015425
2016-03-20    0.020418
2016-03-21    0.020662
2016-03-22    0.021261
2016-03-23    0.018399
2016-03-24    0.019508
2016-03-25    0.018598
2016-03-26    0.016423
2016-03-27    0.015447
2016-03-28    0.020529
2016-03-29    0.021372
2016-03-30    0.024169
2016-03-31    0.023459
2016-04-01    0.022882
2016-04-02    0.024879
2016-04-03    0.024946
2016-04-04    0.024324
2016-04-05    0.126481
2016-04-06    0.225354
2016-04-07    0.134138
Name: last_seen, dtype: float64

Distribuição de valores da colunas last_seen

### Explorando ano de registro dos carros

Iremos explorar se existem valores anormais para os registros dos carros.

In [18]:
autos['registration_year'].describe()

count    45058.000000
mean      2005.063918
std         89.689852
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Os valores máximos e mínimos indicam haver alguns erros, já que não é possível valores de ano igual a 1000 ou 9999.

Consideraremos os anos entre 1950 e 2016 como valores razoáveis

In [19]:
autos = autos[autos['registration_year'].between(1950, 2016)]

In [20]:
autos['registration_year'].value_counts(normalize = True, dropna = False).sort_index().head(50)

1950    0.000023
1951    0.000023
1952    0.000023
1953    0.000023
1954    0.000046
1955    0.000046
1956    0.000092
1957    0.000046
1958    0.000092
1959    0.000139
1960    0.000439
1961    0.000139
1962    0.000092
1963    0.000185
1964    0.000254
1965    0.000393
1966    0.000485
1967    0.000601
1968    0.000578
1969    0.000439
1970    0.000786
1971    0.000578
1972    0.000716
1973    0.000532
1974    0.000555
1975    0.000416
1976    0.000485
1977    0.000485
1978    0.000971
1979    0.000786
1980    0.001757
1981    0.000624
1982    0.000948
1983    0.001156
1984    0.001133
1985    0.001965
1986    0.001502
1987    0.001572
1988    0.002958
1989    0.003582
1990    0.006333
1991    0.006772
1992    0.007026
1993    0.007858
1994    0.011556
1995    0.019853
1996    0.024776
1997    0.034923
1998    0.046756
1999    0.059445
Name: registration_year, dtype: float64

É possível ver que a maioria dos carros são dos anos depois de 1980

### Explorando os dados de marcas

Iremos identificar as marcas que mais vendem.

In [21]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'renault', 'audi', 'sonstige_autos', 'mazda', 'porsche', 'mini',
       'mercedes_benz', 'seat', 'toyota', 'opel', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'jaguar', 'fiat', 'skoda', 'subaru',
       'kia', 'citroen', 'mitsubishi', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [22]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.212148
bmw               0.115076
mercedes_benz     0.102272
opel              0.100423
audi              0.090600
ford              0.065154
renault           0.043936
peugeot           0.029399
fiat              0.023551
seat              0.017912
skoda             0.017172
smart             0.015139
nissan            0.015023
mazda             0.015023
citroen           0.014168
toyota            0.013428
hyundai           0.010169
sonstige_autos    0.009430
mini              0.009384
volvo             0.009384
honda             0.007812
mitsubishi        0.007743
kia               0.007350
alfa_romeo        0.006633
chevrolet         0.006032
suzuki            0.005917
porsche           0.005824
chrysler          0.003606
dacia             0.002843
jeep              0.002450
land_rover        0.002265
daihatsu          0.002173
subaru            0.001988
saab              0.001664
jaguar            0.001618
daewoo            0.001341
rover             0.001248
t

É possível ver que a marca que mais vende é a Volkswagen. Agora iremos escolher as marcas com maior porcentagem de vendas e checar seu preço médio. Escolheremos as marcas que tem uma porcentagem maior que 5% para utilizar.

In [23]:
brand_counts = autos['brand'].value_counts(normalize=True)
best_brands = brand_counts[brand_counts > 0.05].index
print(best_brands)

Index(['volkswagen', 'bmw', 'mercedes_benz', 'opel', 'audi', 'ford'], dtype='object')


Criaremos uma função para que preencherá a média de uma série por marcas

In [24]:
def brands_mean_column(some_brands, data, column):
    """Retorna uma média da coluna escolhida para as marcas selecionadas.

    Args:
        some_brands (str): Marcas escolhidas.
        data (pandas dataframe): DataFrame com os dados.
        column (str): Coluna escolhida.

    Return:
        Dict: Um dicionário com as marcas como palavras chaves
        e os valores médios das marcas.
    """
    brands_mean = {}
    for brand in some_brands:
        brand_chosen = data[data['brand'] == brand]
        mean_column = brand_chosen[column].mean()
        brands_mean[brand] = int(mean_column)
    return brands_mean

In [25]:
brands_mean_prices = brands_mean_column(best_brands, autos, column='price')
print(brands_mean_prices)

{'volkswagen': 5779, 'bmw': 8446, 'mercedes_benz': 8663, 'opel': 3394, 'audi': 9571, 'ford': 4233}


Esses são os preços médios de cada marca. A Audi tem o maior preço médio, junto com Mercedes Benz e BMW, enquanto a Opel e a Ford tem o menor. A Volkswagen tem um valor médio, o que pode explicar ser a marca que mais vende.

### Analisando os dados de preços com os de quilometragem

Para essa análise, iremos criar um dataframe com as séries de preços já calculadas e com uma série de quilometragem. Para isso, iremos criar um dicionário com as quilometragens por marca.

In [26]:
brands_mean_km = brands_mean_column(best_brands, autos, column='odometer_km')
print(brands_mean_km)

{'volkswagen': 128237, 'bmw': 132954, 'mercedes_benz': 131127, 'opel': 128035, 'audi': 128941, 'ford': 123644}


Agora criaremos o dataframe.

In [27]:
bmp_series = pd.Series(brands_mean_prices).sort_values()
bmk_series = pd.Series(brands_mean_km)

df = pd.DataFrame(bmp_series, columns = ['mean_price'])
df['mean_km'] = bmk_series
print(df)

               mean_price  mean_km
opel                 3394   128035
ford                 4233   123644
volkswagen           5779   128237
bmw                  8446   132954
mercedes_benz        8663   131127
audi                 9571   128941


Não há uma influência tão grande dos valores de quilometragem tornando os preços dos carros maiores.